src/volume_rendering/volume_render_layer.ts

```diff
@@ -817,6 +1033,151 @@ void main() {
     gl.disable(WebGL2RenderingContext.CULL_FACE);
     endShader();
     this.vertexIdHelper.disable();
+    if (!needToDrawHistogram) {
+      if (CHECK_PERFORMANCE) {
+        frameRateMonitor.endFrameTimeQuery(gl, ext, query);
+        console.log(frameRateMonitor.getLastFrameTimesInMs(gl, ext));
+      }
+      return;
+    }
+    // Handle histogram drawing
+    let histogramShader: ShaderProgram | null = null;
+    let histogramShaderResult: ParameterizedShaderGetterResult<
+      ShaderControlsBuilderState,
+      VolumeRenderingShaderParameters
+    >;
+    const endHistogramShader = () => {
+      if (histogramShader === null) return;
+      histogramShader.unbindTransferFunctionTextures();
+      if (prevChunkFormat !== null) {
+        prevChunkFormat!.endDrawing(gl, histogramShader);
+      }
+    };
+    const determineNumHistogramInstances = (
+      chunkDataSize: vec3,
+      numHistograms: number,
+    ) => {
+      const maxSamplesInChunk = Math.ceil(
+        chunkDataSize.reduce((a, b) => a * b, 1) / 2.0,
+      );
+      const totalDesiredSamplesInChunk = NUM_HISTOGRAM_SAMPLES / numHistograms;
+      const desiredSamples = Math.min(
+        maxSamplesInChunk,
+        totalDesiredSamplesInChunk,
+      );
+
+      // round to nearest multiple of NUM_HISTOGRAM_SAMPLES_PER_INSTANCE
+      return Math.max(
+        Math.round(desiredSamples / HISTOGRAM_SAMPLES_PER_INSTANCE),
+        1,
+      );
+    };
+
+    prevChunkFormat = null;
+    const { dataType, dataHistogramSpecifications } = this;
+    const outputFramebuffers = dataHistogramSpecifications.getFramebuffers(gl);
+    const count = this.getDataHistogramCount();
+    for (let i = 0; i < count; ++i) {
+      outputFramebuffers[i].bind(256, 1);
+      gl.clearColor(0.0, 0.0, 0.0, 1.0);
+      gl.clear(WebGL2RenderingContext.COLOR_BUFFER_BIT);
+    }
+    const bounds = this.dataHistogramSpecifications.bounds.value;
+    for (let j = 0; j < presentCount; ++j) {
+      newSource = true;
+      const chunkInfo = chunkInfoForHistogram[j];
+      const chunkFormat = chunkInfo.chunkFormat;
+      if (chunkFormat !== prevChunkFormat) {
+        prevChunkFormat = chunkFormat;
+        endHistogramShader();
+        histogramShaderResult = this.histogramShaderGetter({
+          chunkFormat: chunkFormat!,
+        });
+        histogramShader = histogramShaderResult.shader;
+        if (histogramShader !== null) {
+          if (chunkFormat !== null) {
+            chunkFormat.beginDrawing(gl, histogramShader);
+            chunkFormat.beginSource(gl, histogramShader);
+          }
+          histogramShader.bind();
+        } else {
+          return;
```

Comment:
This will break things as the previous framebuffer is not restored, or the timing ended. Should be break I think